reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,286 @@
+## Summary
+### General Motivation
+
+Supporting running Ray applications on spark cluster / databricks runtime,
+Spark is a popular distributed computing framework and is used widely,
+If supporting running Ray applications on spark cluster, user don't need to
+setup a standalone ray cluster and it allows ray workloads and spark workloads
+runs together.
+
+#### Key requirements:
+- Ray application resources (cpu/gpu/memory) allocation must respect spark job resource allocation.
+- Little overhead compared with native Ray cluster.
+- (Optional) Ray UI portal support.
+
+### Should this change be within `ray` or outside?
+
+Yes. For better code maintemance.
+
+## Stewardship
+### Required Reviewers
+
+- @jjyao
+- @ericl
+
+### Shepherd of the Proposal (should be a senior committer)
+
+- @jjyao
+
+## Design and Architecture
+
+### Prerequisites
+
+- The user have an active Spark cluster (apache/spark >= 3.3)
+
+- The user must install Ray packages on every node of the Spark cluster, e.g. via:
+
+  ```
+  pip install ray
+  ```
+
+  If the user's workload requires Ray add-ons, they must also be installed, e.g. via
+
+  ```
+  pip install "ray[debug,dashboard,tune,rllib,serve]"
+  ```
+
+### How to setup Ray cluster over spark cluster ?
+
+The architecture of a Spark cluster is as follows:
+![spark-cluster-overview](https://spark.apache.org/docs/latest/img/cluster-overview.png)
+
+Using Spark's barrier mode, we intend to represent each Ray worker node as a long-running Spark
+task, where the Ray worker node has the same set of resources as the Spark task. Multiple Spark
+tasks can run on a single Spark worker node, which means that multiple Ray worker nodes can run
```

Comment:
> Multiple Spark
tasks can run on a single Spark worker node, which means that multiple Ray worker nodes can run
on a single Spark worker node.

Is this a requirement? Is it possible to promote 1 node <> 1 ray instance as the best practice? It technically works, but Ray may not play well with this setup sometimes (more than one runtime instance per node). For example, many dashboard features rely on node IP. There are some other hiccups we don't expect (like we rely on some files inside /tmp/ray).

Reply:
> Is it possible to promote 1 node <> 1 ray instance as the best practice?

yes. this is the best practice.

We can config spark task shape to make only one spark task slots on each spark worker node, so that it will satisfy the best practice rule.